Libraries/LibCore/Account.cpp

```diff
@@ -0,0 +1,202 @@
+{
+    struct passwd* pwd = nullptr;
+    errno = 0;
+    pwd = getpwuid(uid);
+    if (!pwd) {
+        if (errno == 0)
+            return String("No such user");
+
+        return String(strerror(errno));
+    }
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+bool Account::authenticate(const char* password) const
+{
+    // An empty passwd field indicates that no password is required to log in.
+    if (m_password_hash.is_empty())
+        return true;
+
+    // FIXME: Use crypt_r if it can be built in lagom.
+    char* hash = crypt(password, m_password_hash.characters());
+    return hash != nullptr && strcmp(hash, m_password_hash.characters()) == 0;
+}
+
+bool Account::login() const
+{
+    if (setgroups(m_extra_gids.size(), m_extra_gids.data()) < 0)
+        return false;
+
+    if (setgid(m_gid) < 0)
+        return false;
+
+    if (setuid(m_uid) < 0)
+        return false;
+
+    return true;
+}
+
+void Account::set_password(const char* password)
+{
+    m_password_hash = crypt(password, get_salt().characters());
+}
+
+void Account::set_password_enabled(bool enabled)
+{
+    if (enabled && m_password_hash != "" && m_password_hash[0] == '!') {
+        m_password_hash = m_password_hash.substring(1, m_password_hash.length() - 1);
+    } else if (!enabled && (m_password_hash == "" || m_password_hash[0] != '!')) {
+        StringBuilder builder;
+        builder.append('!');
+        builder.append(m_password_hash);
+        m_password_hash = builder.build();
+    }
+}
+
+void Account::delete_password()
+{
+    m_password_hash = "";
+}
+
+Account::Account(struct passwd* pwd, Vector<gid_t> extra_gids)
+    : m_username(pwd->pw_name)
+    , m_password_hash(pwd->pw_passwd)
+    , m_uid(pwd->pw_uid)
+    , m_gid(pwd->pw_gid)
+    , m_gecos(pwd->pw_gecos)
+    , m_home_directory(pwd->pw_dir)
+    , m_shell(pwd->pw_shell)
+    , m_extra_gids(extra_gids)
+{
+}
+
+bool Account::sync()
+{
+    StringBuilder new_passwd_file;
+
+    setpwent();
+
+    struct passwd* p;
+    while ((p = getpwent())) {
+        if (p->pw_uid == m_uid) {
+            new_passwd_file.appendf("%s:%s:%u:%u:%s:%s:%s\n",
+                m_username.characters(),
+                m_password_hash.characters(),
+                m_uid, m_gid,
+                m_gecos.characters(),
+                m_home_directory.characters(),
+                m_shell.characters());
+
+        } else {
+            new_passwd_file.appendf("%s:%s:%u:%u:%s:%s:%s\n",
+                p->pw_name, p->pw_passwd, p->pw_uid,
+                p->pw_gid, p->pw_gecos, p->pw_dir,
+                p->pw_shell);
+        }
+    }
+    endpwent();
+
+    String contents = new_passwd_file.build();
+
+    FILE* passwd_file = fopen("/etc/passwd", "w");
+    if (!passwd_file)
+        return false;
+
+    fwrite(contents.characters(), 1, contents.length(), passwd_file);
+    if (ferror(passwd_file)) {
+        // fclose sets errno if write fails.
```

Comment:
But what if the `fwrite()` above failed (say, with `ENOSPC`), but `fclose()` succeeds?

I mean, it tires to flush, so it'll probably get the same error again, but I wouldn't rely on that.

Reply:
> Would you prefered that I set `errno = ferror()`, or return an error number?

Just make sure to do the same thing no matter which exit path this function takes 🙂 I think you're already setting `errno` in other places (or calling functions that set it for you), so I'd set `errno = ferror()` here.